crates/uv-interpreter/src/discovery.rs

```diff
@@ -0,0 +1,1134 @@
+use itertools::Itertools;
+use thiserror::Error;
+use tracing::{debug, instrument, trace};
+use uv_cache::Cache;
+use uv_fs::Simplified;
+use uv_warnings::warn_user_once;
+use which::which;
+
+use crate::implementation::ImplementationName;
+use crate::managed::toolchains_for_current_platform;
+use crate::py_launcher::py_list_paths;
+use crate::virtualenv::{
+    virtualenv_from_env, virtualenv_from_working_dir, virtualenv_python_executable,
+};
+use crate::{Interpreter, PythonVersion};
+use std::borrow::Cow;
+
+use std::collections::HashSet;
+use std::fmt::{self, Formatter};
+use std::num::ParseIntError;
+use std::{env, io};
+use std::{path::Path, path::PathBuf, str::FromStr};
+
+/// A request to find a Python interpreter.
+///
+/// See [`InterpreterRequest::from_str`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum InterpreterRequest {
+    /// A Python version without an implementation name e.g. `3.10`
+    Version(VersionRequest),
+    /// A path to a directory containing a Python installation, e.g. `.venv`
+    Directory(PathBuf),
+    /// A path to a Python executable e.g. `~/bin/python`
+    File(PathBuf),
+    /// The name of a Python executable (i.e. for lookup in the PATH) e.g. `foopython3`
+    ExecutableName(String),
+    /// A Python implementation without a version e.g. `pypy`
+    Implementation(ImplementationName),
+    /// A Python implementation name and version e.g. `pypy3.8`
+    ImplementationVersion(ImplementationName, VersionRequest),
+}
+
+/// The sources to consider when finding a Python interpreter.
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub enum SourceSelector {
+    #[default]
+    All,
+    Some(HashSet<InterpreterSource>),
+}
+
+/// A Python interpreter version request.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub enum VersionRequest {
+    #[default]
+    Default,
+    Major(u8),
+    MajorMinor(u8, u8),
+    MajorMinorPatch(u8, u8, u8),
+}
+
+/// The policy for discovery of "system" Python interpreters.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum SystemPython {
+    // Do not allow a system Python
+    #[default]
+    Disallowed,
+    // Allow a system Python to be used if no vritual environment is active.
+    Allowed,
+    // Prefer a system Python over virtual environments.
+    Preferred,
```

Comment:
Today, doesn't `--today` _ignore_ (e.g.) discovering a `.venv` in the current directory? `Preferred` makes it sound like that's still allowed as a fallback. Is it a behavior change?

Reply:
Looking at the code, I think it _does_ ignore the `.venv` in the current directory. So maybe this should be called `Required`...?

Reply:
Yeah you are correct